sdks/python/apache_beam/runners/dataflow/internal/apiclient.py

```diff
@@ -740,6 +720,28 @@ def _apply_sdk_environment_overrides(
       new_payload.container_image = new_container_image
       environment.payload = new_payload.SerializeToString()
 
+    # De-dup environments that use Java SDK by Docker container image since
+    # currently running multiple Java SDK Harnesses with Dataflow could result
+    # in dependency conflicts.
+    # TODDO(BEAM-9455): remove following restriction when Dataflow supports
+    # environment specific dependency provisioning.
+    container_url_to_env_map = dict()
+    container_url_to_env_id_map = dict()
+    for transform in proto_pipeline.components.transforms.values():
+      environment_id = transform.environment_id
+      if not environment_id:
+        continue
+      environment = proto_pipeline.components.environments[environment_id]
+      docker_payload = proto_utils.parse_Bytes(
+          environment.payload, beam_runner_api_pb2.DockerPayload)
+      image = docker_payload.container_image
+      if is_beam_java_container_name(image):
```

Comment:
Is it possible to do this rewrite later (before submitting the v1beta3 job)? I am seeing that this code is will be called before we translate beam pipeline into v1beta3 Steps. It will be difficult to support resource hints for xlang with this rewrite. You also need to do the same in portable job submission codepath in DF service. I wonder if doing this rewrite on cloud workflow proto level (service change) would be a better path.

Reply:
Discussed this offline, I am ok to proceed as is with understanding that resource hint support will not work for xlang until we revert this workaround and fix the rootcause.
cc: @robertwb

Reply:
Thanks, we had an offline discussion and seems like this PR does not add any additional restrictions related to resource hints implementation given that Dataflow cannot support multiple Java Environments (with multiple Java SDK Harnesses) till we fix Dataflow Dependencies issue anyways (https://issues.apache.org/jira/browse/BEAM-9455).

Reply:
Ah, sorry didn't notice your comment above. Thanks.